x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SparseScoredTermsMapper.java

```diff
@@ -0,0 +1,288 @@
+                        "failed to parse field [%s] of type [%s], [terms] and [scores] must have equal length; found [%d] and [%d]",
+                        fieldType().name(),
+                        CONTENT_TYPE,
+                        doc.terms.size(),
+                        doc.scores.size()
+                    )
+                );
+            }
+            int i = 0;
+            for (Integer score : doc.scores) {
+                if (score < 0) {
+                    throw new MapperParsingException(
+                        Strings.format(
+                            "failed to parse field [%s] of type [%s], [scores] must be non-negative; found [%d] at position [%d]",
+                            fieldType().name(),
+                            CONTENT_TYPE,
+                            score,
+                            i
+                        )
+                    );
+                }
+                ++i;
+            }
+            if (doc.terms.size() > Sets.newHashSet(doc.terms).size()) {
+                throw new MapperParsingException(
+                    Strings.format(
+                        "failed to parse field [%s] of type [%s]; requires unique items in [terms]",
+                        fieldType().name(),
+                        CONTENT_TYPE
+                    )
+                );
+            }
+            context.doc()
+                .addWithKey(fieldType().name(), new Field(name(), new TermCountStream(doc.terms, doc.scores), Defaults.FIELD_TYPE));
+            context.addToFieldNames(fieldType().name());
+        } catch (Exception ex) {
+            if (ignoreMalformed.value() == false) {
+                if (ex instanceof MapperParsingException) {
+                    throw ex;
+                }
+                throw new MapperParsingException("failed to parse field [{}] of type [{}]", ex, fieldType().name(), fieldType().typeName());
+            }
+
+            if (subParser != null) {
+                // close the subParser so we advance to the end of the object
+                subParser.close();
+            }
+            context.addIgnoredField(fieldType().name());
+        }
+        context.path().remove();
+    }
+
+    @Override
+    public Map<String, NamedAnalyzer> indexAnalyzers() {
+        return Map.of(mappedFieldType.name(), Lucene.KEYWORD_ANALYZER);
+    }
+
+    @Override
+    public FieldMapper.Builder getMergeBuilder() {
+        return new Builder(simpleName(), ignoreMalformedByDefault).init(this);
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    public boolean ignoreMalformed() {
+        return ignoreMalformed.value();
+    }
+
+    private static class TermCountStream extends TokenStream {
+
+        private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+        private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
```

Comment:
Is there any point setting the OffsetAttribute here? We're not indexing positions or offsets...